### ISSUE TITLE
Size of the saved model checkpoint after extending pretraining using FSDP is larger than that of the base model.

### ISSUE THREAD

**Issue by dineshkh** [huggingface/accelerate#1958]
I am do extended pretraining of a 20B model using FSDP. I observe that the size of model saved after extended pretraining is larger than the base model. This is how I am saving the model. 

```
       unwrapped_model = accelerator.unwrap_model(model)   
        state_dict = None
        if os.environ.get("ACCELERATE_USE_FSDP") == "true":
            logger.info(f'using ACCELERATE plugin')
            full_state_dict_config = FullStateDictConfig(offload_to_cpu=True, rank0_only=True)
            with FSDP.state_dict_type(unwrapped_model,
                                      StateDictType.FULL_STATE_DICT, full_state_dict_config):
                state_dict = accelerator.get_state_dict(unwrapped_model)
        if accelerator.is_main_process:
            unwrapped_model.save_pretrained(
                output_dir,
                is_main_process=accelerator.is_main_process,
                save_function=accelerator.save,
                state_dict=state_dict,
                safe_serialization=FLAGS.safe_serialization
            )
```
I am using `safetensors` for saving the model. 

Size of base model on disk:

<img width="662" alt="image" src="https://github.com/huggingface/accelerate/assets/14121108/9e1840ad-624c-4d25-a7e3-42195271919e">

Size of extended pretrained model on disk:

<img width="803" alt="image" src="https://github.com/huggingface/accelerate/assets/14121108/125584c2-198b-4cbc-8737-885d2f8abb10">

If you see size of last shard has increased from `1359201312` to `2567185552`
Any reason of this behavior ?

Following is my FSDP configuration:
```
compute_environment: LOCAL_MACHINE
distributed_type: FSDP
fsdp_config:
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_backward_prefetch_policy: NO_PREFETCH
  fsdp_offload_params: false
  fsdp_sharding_strategy: 1
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_transformer_layer_cls_to_wrap: GPTBigCodeBlock
machine_rank: 0
main_process_ip: 172.168.12.19
main_process_port: 55930
main_training_function: main
mixed_precision: 'bf16'
num_machines: 10
num_processes: 80
use_cpu: false
```

I am using `accelerate==0.20.3`. My transformers and torch version are following:

`transformers              4.29.0                   pypi_0    pypi`
`torch                     2.0.1                    pypi_0    pypi`

**Comment by LysandreJik** [comment]
cc @pacman100

**Comment by pacman100** [comment]
Hello @dineshkh, could you please try the latest released version of Accelerate and let us know if the issue still persists.

**Comment by dineshkh** [comment]
@pacman100 I have tried with `accelerate==0.22.0` and but still saved model checkpoint size is larger compared to the base model. 

Also, with accelerate==0.23.0 my code is running for 2 steps and giving OOM error. The same code is working when I was using `accelerate==0.20.3` or `accelerate==0.22.0`. I am using `gradient checkpointing` and `flash attention v2` via monkey patch provided by you.

```
    if FLAGS.gradient_checkpointing:
        model.gradient_checkpointing_enable()
```

**Comment by dineshkh** [comment]
@pacman100, @LysandreJik  any update on this ?

**Comment by pacman100** [comment]
Hello, I did the following:
1. Code `issue_1958.py`:
```
from accelerate import Accelerator
import torch
from transformers import AutoModelForCausalLM, get_linear_schedule_with_warmup
from torch.distributed.fsdp import (
    FullyShardedDataParallel as FSDP,
    StateDictType,
    FullStateDictConfig,
)


def main():
    model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-chat-hf")
    accelerator = Accelerator()
    model.save_pretrained("delete_me_base", is_main_process=accelerator.is_main_process)
    model = accelerator.prepare(model)
    optimizer = torch.optim.AdamW(model.parameters(), lr=5e-5)
    lr_scheduler = get_linear_schedule_with_warmup(
        optimizer=optimizer,
        num_warmup_steps=0,
        num_training_steps=1000,
    )
    optimizer, lr_scheduler = accelerator.prepare(optimizer, lr_scheduler)

    full_state_dict_config = FullStateDictConfig(offload_to_cpu=True, rank0_only=True)
    with FSDP.state_dict_type(
        model, StateDictType.FULL_STATE_DICT, full_state_dict_config
    ):
        state_dict = accelerator.get_state_dict(model, unwrap=False)

    unwrapped_model = accelerator.unwrap_model(model)
    unwrapped_model.save_pretrained(
        "delete_me",
        is_main_process=accelerator.is_main_process,
        save_function=accelerator.save,
        state_dict=state_dict,
    )
    print("Done saving model")


if __name__ == "__main__":
    main()
```

2. Accelerate config `accelerate config --config_file issue_1958.yaml`
```
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: FSDP
downcast_bf16: 'no'
fsdp_config:
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_backward_prefetch_policy: BACKWARD_PRE
  fsdp_cpu_ram_efficient_loading: true
  fsdp_forward_prefetch: false
  fsdp_offload_params: false
  fsdp_sharding_strategy: 1
  fsdp_state_dict_type: SHARDED_STATE_DICT
  fsdp_sync_module_states: true
  fsdp_use_orig_params: true
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 4
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

3.  run `accelerate launch --config_file issue_1958.yaml issue_1958.py `:
```
Loading checkpoint shards: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.53s/it]
Loading checkpoint shards: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.53s/it]
Detected kernel version 5.4.0, which is below the recommended minimum of 5.5.0; this can cause the process to hang. It is recommended to upgrade the kernel to the minimum version or higher.
Loading checkpoint shards: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.90s/it]
Loading checkpoint shards: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:03<00:00,  1.90s/it]
Done saving model
Done saving model
Done saving model
Done saving model
```


4. Inspecting base model weights `ls -lh delete_me_base/`:
```
total 26G
-rw-rw-r-- 1 sourab sourab  694 nov.  21 08:52 config.json
-rw-rw-r-- 1 sourab sourab  188 nov.  21 08:52 generation_config.json
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:52 model-00001-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:52 model-00002-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:52 model-00003-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:52 model-00004-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:52 model-00005-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 2,6G nov.  21 08:53 model-00006-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab  24K nov.  21 08:53 model.safetensors.index.json
```

5. Inspecting fsdp saved model `ls -lh delete_me`:
```
total 26G
-rw-rw-r-- 1 sourab sourab  694 nov.  21 08:53 config.json
-rw-rw-r-- 1 sourab sourab  188 nov.  21 08:53 generation_config.json
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:53 model-00001-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:53 model-00002-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:53 model-00003-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:53 model-00004-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 4,6G nov.  21 08:53 model-00005-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab 2,6G nov.  21 08:54 model-00006-of-00006.safetensors
-rw-rw-r-- 1 sourab sourab   48 nov.  21 08:53 model.safetensors
-rw-rw-r-- 1 sourab sourab  24K nov.  21 08:54 model.safetensors.index.json
```

Therefore, the sizes of the checkpoints are same. It isn't increasing due to FSDP.
